### PR TITLE
Wrap plot3d

### LINKED ISSUE #467: Wrapper for plot3d
**Description of the desired feature**

Implement [`plot3d`](https://docs.generic-mapping-tools.org/latest/plot3d.html) which "Plot lines, polygons, and symbols in 3-D". Basically a 3D version of ['plot`](https://docs.generic-mapping-tools.org/latest/plot.html). Below is an example of such a 3D plot from the GMT gallery:

Example 8 using plot3d from GMT gallery

This would build upon the [`grdview`](https://www.pygmt.org/latest/api/generated/pygmt.Figure.grdview.html) function I last wrapped in #330, extending it to vector symbols. For example, we could plot the change in height of a DEM surface (h) along a transect line (x), over time (t)! Will be keen to get this done for some of my current satellite altimetry work :grin:

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by weiji14**
Wrapping the `plot3d` function! Original GMT `plot3d` function can be found at https://docs.generic-mapping-tools.org/latest/plot3d.html. Current implementation is mostly copy-modify-pasted from `plot`, including the tests. Also added extra aliases to the `basemap` function to make it 3D capable.

Live documentation preview of `plot3d` API is at https://pygmt-git-mapping-plot3d.gmt.vercel.app/api/generated/pygmt.Figure.plot3d.html. Gallery example is at https://pygmt-git-mapping-plot3d.gmt.vercel.app/gallery/plot/scatter3d.html

**Description of proposed changes**

This is the gallery example of a 3D plot with the [Iris flower dataset](https://en.wikipedia.org/wiki/Iris_flower_data_set), inspired by the example at https://scikit-learn.org/stable/auto_examples/cluster/plot_cluster_iris.html :smile:

3D scatter plot of Iris flower dataset made using PyGMT plot3d

Here's another example with a bit of a geographic curve:

```python
fig = Figure()
fig.plot3d(
    x=data[:, 0],
    y=data[:, 1],
    z=data[:, 2],
    zscale=5,
    perspective=[225, 30],
    region=[10, 70, -5, 10, 0, 1],
    projection="R270/4i",
    style="s1c",
    color="green",
    frame=["ag", "zag"],
)
fig.savefig("example_plot3d.png")
fig.show()
```
produces:
example_plot3d

Alternatively, we could follow GMT.jl's example at https://www.generic-mapping-tools.org/GMT.jl/latest/plot/ and just have one `plot` function, which would handle 3D plotting if the `z` dimension is provided. Thoughts?

Fixes #467

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [x] If adding new functionality, add an example to docstrings or tutorials.

### CODE REVIEW

File: pygmt/base_plotting.py [hunk 1]
```diff
@@ -816,6 +816,175 @@ def plot(self, x=None, y=None, data=None, sizes=None, direction=None, **kwargs):
...
+        C="cmap",
+        D="offset",
+        G="color",
+        I="intensity",
+        J="projection",
+        Jz="zscale",
+        JZ="zsize",
+        L="close",
+        N="no_clip",
+        Q="no_sort",
```
Comment (weiji14) [inline pygmt/base_plotting.py:832]: Note that error_bar(E) is available in `plot`, but not in `plot3d`. Also, no_sort ([Q](https://docs.generic-mapping-tools.org/latest/plot3d.html#q)) is a new alias specific to `plot3d` only (not found in GMT.jl either)!

File: pygmt/base_plotting.py [hunk 2]
```diff
@@ -816,6 +816,175 @@ def plot(self, x=None, y=None, data=None, sizes=None, direction=None, **kwargs):
...
+        R="region",
+        S="style",
+        V="verbose",
+        W="pen",
+        X="xshift",
+        Y="yshift",
+        i="columns",
```
Comment (weiji14) [inline pygmt/base_plotting.py:839]: I've aliased columns (i) here to make it consistent with `plot`. Let me know if this is a good idea, or if we should remove this alias as suggested in the discussion in https://github.com/GenericMappingTools/pygmt/pull/666#discussion_r511538159.

**Review by weiji14 — COMMENTED** [review]
Alright, quite a big PR to review! The test_plot3d.py file is huge, but it is almost an identical copy of test_plot.py, just made for the 3D world and uses `@check_figures_equal` so no baseline images are stored. For reference, the baseline images can be viewed at commit 8391169134331bdd52fdfc9ee75acc666b12f525. Here's an example:



I'll try and create a gallery example too :wink:

File: pygmt/base_plotting.py (hunk 2 above)
Comment (seisman) [reply pygmt/base_plotting.py:840]: Let's not add it the plot3d.

File: pygmt/base_plotting.py [hunk 3]
```diff
@@ -816,6 +816,185 @@ def plot(self, x=None, y=None, data=None, sizes=None, direction=None, **kwargs):
...
+                )
+            extra_arrays.append(kwargs["G"])
+            del kwargs["G"]
+        if sizes is not None:
+            if kind != "vectors":
+                raise GMTInvalidInput(
+                    "Can't use arrays for sizes if data is matrix or file."
+                )
+            extra_arrays.append(sizes)
+
```
Comment (seisman) [inline pygmt/base_plotting.py:982]: Do you want to add the *record-by-record* transparency feature (#615) following #626? It's OK if you want to do it in a separate PR.

Comment (weiji14) [reply pygmt/base_plotting.py:982]: Yep, I can add it. Forgot to include this as I copied the `plot` code from ages ago (pre v0.2.0)!

File: pygmt/base_plotting.py (hunk 2 above)
Comment (seisman) [reply pygmt/base_plotting.py:840]: I just saw that you also used *columns* in the tests. Perhaps it's OK to add it, and then break the backward compatibility with **plot()**.

Comment (weiji14) [reply pygmt/base_plotting.py:840]: Sounds good.

File: pygmt/base_plotting.py (hunk 3 above)
Comment (weiji14) [reply pygmt/base_plotting.py:982]: Alright, done! Added some tests too and made sure to use 3-D symbols, here's an example:

test_plot3d_varying_transparency

https://github.com/GenericMappingTools/pygmt/blob/cdcd25927b624ed548ca4e38319f5406680df20f/pygmt/tests/test_plot3d.py#L371-L383

File: pygmt/tests/test_plot3d.py [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+    fig_ref, fig_test = Figure(), Figure()
+    fig_ref.plot3d(
+        data=POINTS_DATA,
+        Jz=5,
+        p="225/30",
+        R="/".join(map(str, region)),
+        J="M10i",
+        B=["af", "zaf"],
+        G="+z",
+        i="0,1,2,2,2+s1",
```
Comment (weiji14) [inline pygmt/tests/test_plot3d.py:289]: Ok, been trying (unsuccessfully) to see why 3 Windows tests are failing. Specifically `test_plot3d_sizes`, `test_plot3d_colors_sizes`, `test_plot3d_colors_sizes_proj`. I can't get my university Windows box to run `pytest` somehow today (keep getting a `pygmt.exceptions.GMTCLibError: Module 'psconvert' failed with status code 78:
E           psconvert [ERROR]: System call [@gswin64c -q ... returned error 1]` message). So no idea what the image difference looks like other than a vague GMTImageComparisonFailure message.

```
E               pygmt.exceptions.GMTImageComparisonFailure: images not close (RMS 28.107):
E               	result_images\test_plot3d_colors_sizes_proj[png].png
E               	result_images\test_plot3d_colors_sizes_proj[png]-expected.png
```


**However**, I think it might have to do with using the column scaling `i=2+s1`, since the 3 test failures have these. Is this something Windows is particularly sensitive to?

File: pygmt/base_plotting.py (hunk 3 above)
Comment (seisman) [reply pygmt/base_plotting.py:982]: Great!

File: pygmt/tests/test_plot3d.py (hunk 4 above)
Comment (seisman) [reply pygmt/tests/test_plot3d.py:289]: We should upload the testing directory (`tmp-test-dir-with-unique-name`) as an artifact (https://github.com/actions/upload-artifact) if the testing fails.

Comment (weiji14) [reply pygmt/tests/test_plot3d.py:289]: Yep, it's to do with the `i="2+s1"` scaling problem. I can try to refactor the test a bit to make it work cross-platform, but this bug might hit some of our users at a later date.

| expected  | actual  | diff  |
|---|---|---|
| test_plot3d_colors_sizes_proj png | test_plot3d_colors_sizes_proj png -expected| test_plot3d_colors_sizes_proj png -failed-diff |

P.S. Will cherry-pick commit 413ff77c98facb8370cd5bf8e53c6dfe54bd7848 into a separate Pull Request.

Comment (seisman) [reply pygmt/tests/test_plot3d.py:289]: > P.S. Will cherry-pick commit [413ff77](https://github.com/GenericMappingTools/pygmt/commit/413ff77c98facb8370cd5bf8e53c6dfe54bd7848) into a separate Pull Request.

There are 3 failing jobs, but it seems I only see one artifact. Perhaps in 413ff77c98facb8370cd5bf8e53c6dfe54bd7848 we should use unique names for artifacts from different jobs?

Comment (weiji14) [reply pygmt/tests/test_plot3d.py:289]: > There are 3 failing jobs, but it seems I only see one artifact. Perhaps in [413ff77](https://github.com/GenericMappingTools/pygmt/commit/413ff77c98facb8370cd5bf8e53c6dfe54bd7848) we should use unique names for artifacts from different jobs?

Isn't it a separate artifact under each job (i.e. WIndows Py36/Py37/Py38?), I didn't look too closely at the files but I assumed it would be separate.

Comment (seisman) [reply pygmt/tests/test_plot3d.py:289]: I only see one artifact in this page: https://github.com/GenericMappingTools/pygmt/actions/runs/330476822

Comment (weiji14) [reply pygmt/tests/test_plot3d.py:289]: Anyways, I think there might be a bug (or feature?) here somewhere, as using `i="0,1,2,2"` with `-Suc` doesn't change the unit to **c**entimetres, but keeps on using inches as the default unit. The documentation at https://docs.generic-mapping-tools.org/latest/plot3d.html#s confuses me, especially with the mention of `PROJ_LENGTH_UNIT`.

```
gmt begin temp png
  gmt plot3d pygmt/tests/data/points.txt -Jz5 -p225/30 -i0,1,2,2 -R10/70/-5/10/0/1 -JX4i -Suc -Gblue -Baf -Bzaf
gmt end show
```

produces cubes sized in inches instead of centimetres

temp

Comment (seisman) [reply pygmt/tests/test_plot3d.py:289]: I think you're right. **-Suc** and **-Sui** give the same result. It may be a GMT bug.

Comment (weiji14) [reply pygmt/tests/test_plot3d.py:289]: Great, time to open up a new GMT bug report then. I've also made a new PR on uploading artifacts opened at #675 so discussion can continue there.

In the meantime, I'll change these tests to use inches which will hopefully make things pass on Windows.

Comment (seisman) [reply pygmt/tests/test_plot3d.py:289]: Please report the bug to GMT, and also add a few comments in the tests explaining why we use inches.

Comment (weiji14) [reply pygmt/tests/test_plot3d.py:289]: Cool, all tests green on commit 878a07351afbf17928e0293dfb532464caf1e94a when using inches, even with column scaling `i="2+s0.5". I'll revert 413ff77c98facb8370cd5bf8e53c6dfe54bd7848 in this PR (since #675 is opened) and re-request a review to trigger the GMT Latest tests.

Comment (weiji14) [reply pygmt/tests/test_plot3d.py:293]: > Please report the bug to GMT, and also add a few comments in the tests explaining why we use inches.

Ok, bug reported at https://github.com/GenericMappingTools/gmt/issues/4386, and I've added notes to the 3 unit tests on why inches are used instead of centimetres at commit 256a0a9b1a9333827ec9a9e8b36e7b45785d4a19.

File: examples/gallery/plot/scatter3d.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+# The below example will return a numpy array like [0., 3., 4., 8., 1., 7.]
+region = pygmt.info(
+    table=df[["petal_width", "sepal_length", "petal_length"]],  # x, y, z columns
+    per_column=True,  # report output as a numpy array
+    spacing="1/2/0.5",  # rounds x, y and z intervals by 1, 2 and 0.5 respectively
+)
+
+# Make our 3D scatter plot, coloring each of the 3 species differently
+fig = pygmt.Figure()
+pygmt.makecpt(cmap="cubhelix", series=(0, 3, 1))
```
Comment (seisman) [inline examples/gallery/plot/scatter3d.py:33]: This `makecpt` command produces a discrete CPT (see the figure below). It doesn't make sense for the categorical data here. 
image. 

Instead, we should use `makecpt -F+c`, which produces categorical CPT.

image

But the colorbar annotations don't make sense to me. Perhaps another GMT bug.

Note: the gridlines are not shown in GMT 6.1.1. It's a new future in GMT master (see https://github.com/GenericMappingTools/gmt/pull/3993), but the gridlines in XZ and YZ planes may also have bugs. Will report to GMT later.

File: doc/api/index.rst [hunk 6]
```diff
@@ -26,6 +26,7 @@ Plotting data and laying out the map:
     Figure.coast
     Figure.colorbar
     Figure.plot
+    Figure.plot3d
```
[the remark this answers is not in the document]
Comment (weiji14) [reply doc/api/index.rst:29]: Maybe just for the `Figure.*` methods, but that's for a separate PR.

File: examples/gallery/plot/scatter3d.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+# Load sample iris data, and convert 'species' column to categorical dtype
+df = pd.read_csv("https://github.com/mwaskom/seaborn-data/raw/master/iris.csv")
+df["species"] = df.species.astype(dtype="category")
+
+# Use pygmt.info to get region bounds (xmin, xmax, ymin, ymax, zmin, zmax)
+# The below example will return a numpy array like [0., 3., 4., 8., 1., 7.]
+region = pygmt.info(
+    table=df[["petal_width", "sepal_length", "petal_length"]],  # x, y, z columns
+    per_column=True,  # report output as a numpy array
+    spacing="1/2/0.5",  # rounds x, y and z intervals by 1, 2 and 0.5 respectively
```
[the remark this answers is not in the document]
Comment (weiji14) [reply examples/gallery/plot/scatter3d.py:28]: Yeah, forgot to handle that in https://github.com/GenericMappingTools/pygmt/pull/575. Should handle that (list inputs) when we complete the documentation for `info`.

File: examples/gallery/plot/scatter3d.py (hunk 5 above)
Comment (seisman) [reply examples/gallery/plot/scatter3d.py:33]: Already reported to upstream https://github.com/GenericMappingTools/gmt/issues/4387 and https://github.com/GenericMappingTools/gmt/issues/4388.

File: doc/api/index.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (seisman) [reply doc/api/index.rst:29]: Yes.

File: examples/gallery/plot/scatter3d.py (hunk 5 above)
Comment (weiji14) [reply examples/gallery/plot/scatter3d.py:33]: > Instead, we should use makecpt -F+c, which produces categorical CPT.
> But the colorbar annotations don't make sense to me. Perhaps another GMT bug.

Ah yes, forgot about [`-F+c`](https://docs.generic-mapping-tools.org/latest/makecpt.html#f)! Actually for the Iris species categories are nominal (discrete) rather than ordinal (i.e. ranked 1st, 2nd, 3rd), so a legend would make more sense here than a colorbar. Another thought I had was to label each species using `text` plotted in 3D, but we haven't implemented that 'feature' yet.

So for now, I think we can just use `-F+c` here and ignore the colorbar, though we should add another categorical gallery example to close #244? I'd like to avoid adding a legend since it requires a for-loop to make the labels (and this complicates the example).

The wall frames look good though, too bad we need to wait for GMT 6.2.0! But good that we found two bugs to report to upstream GMT :laughing:

Comment (weiji14) [reply examples/gallery/plot/scatter3d.py:33]: Actually, `makecpt -F` is not aliased yet. Do we want to do that here or in a separate PR?

Comment (weiji14) [reply examples/gallery/plot/scatter3d.py]: Alright, with #676 merged, I've updated the code to use `pygmt.makecpt(..., color_model="+c", ...)`

https://github.com/GenericMappingTools/pygmt/blob/40f7120cc20dc96aac8ab8d431dfd4b124274b6e/examples/gallery/plot/scatter3d.py#L34

In the future when we bump the minimum GMT version to 6.2.0 which would include the new feature at https://github.com/GenericMappingTools/gmt/pull/4390, we can do the following:

```pygmt
pygmt.makecpt(cmap="cubhelix", color_model="+cSetosa,Versicolor,Virginica", series=(0, 3, 1))
fig.plot3d(...)
fig.colorbar()
fig.show()
```

which would produce a nice colorbar ;)

File: examples/gallery/plot/scatter3d.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+"""
+3D Scatter plots
+----------------
+
+The :meth:`pygmt.Figure.plot3d` method can be used to plot symbols in 3D.
+In the example below, we show how the
+`Iris flower dataset <https://en.wikipedia.org/wiki/Iris_flower_data_set/>`__.
```
Comment (seisman) [inline examples/gallery/plot/scatter3d.py:7]:
```suggestion
`Iris flower dataset <https://en.wikipedia.org/wiki/Iris_flower_data_set/>`__
```

File: examples/gallery/plot/scatter3d.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+fig.plot3d(
+    x=df.petal_width,
+    y=df.sepal_length,
+    z=df.petal_length,
+    sizes=0.1 * df.sepal_width,  # Vary each symbol size according to a data column
+    color=df.species.cat.codes.astype(int),  # Points colored by categorical number code
+    cmap=True,  # Use colormap created by makecpt
+    region=region,  # (xmin, xmax, ymin, ymax, zmin, zmax)
+    frame=[
+        "WsNeZ3",  # z axis label positioned on 3rd corner
```
Comment (seisman) [inline examples/gallery/plot/scatter3d.py:44]: What about adding a title here?

Comment (weiji14) [reply examples/gallery/plot/scatter3d.py:44]: The title looks like it is placed too high without the gridlines at the back, doesn't look very nice:

tmp_iris

Comment (seisman) [reply examples/gallery/plot/scatter3d.py:44]: Agree

**Review by seisman — APPROVED** [review]

**Comment by weiji14** [comment]
Thanks for the thorough review @seisman!
